runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -264,29 +276,35 @@ public void write(IToken token, Object data) throws OverwriteException {
      * @return A result, which be cached.
      */
     public Map<Long, ILogData> read(Iterable<Long> addresses) {
-        Map<Long, ILogData> addressesMap;
         if (!runtime.getParameters().isCacheDisabled()) {
-            try {
-                addressesMap = readCache.getAll(addresses);
-            } catch (ExecutionException | UncheckedExecutionException e) {
-                Throwable cause = e.getCause();
-                if (cause instanceof RuntimeException) {
-                    throw (RuntimeException) cause;
+            Map<Long, ILogData> result = new HashMap<>();
+            Set<Long> addressesToFetch = new HashSet<>();
+
+            for (Long address : addresses) {
+                ILogData val = readCache.getIfPresent(address);
+                if (val == null) {
+                    addressesToFetch.add(address);
                 } else {
-                    throw new RuntimeException(cause);
+                    result.put(address, val);
                 }
             }
-        } else {
-            addressesMap = this.cacheFetch(addresses);
-        }
 
-        for (ILogData logData : addressesMap.values()) {
-            if (logData.isTrimmed()) {
-                throw new TrimmedException();
+            // At this point we computed a subset of the addresses that
+            // resulted in a cache miss and need to be fetched
+            if (!addressesToFetch.isEmpty()) {
+                Map<Long, ILogData> fetchedAddresses = fetchAll(addressesToFetch);
+                for (Map.Entry<Long, ILogData> entry : fetchedAddresses.entrySet()) {
+                    // After fetching a value, we need to insert it in the cache.
+                    // Note that based on code inspection it seems like operations
+                    // on the cache's map view are reflected in the cache's statistics.
+                    result.put(entry.getKey(), readCache.asMap()
+                            .computeIfAbsent(entry.getKey(), (k) -> entry.getValue()));
```

Comment:
The operation seems not atomic.
What if during your fetching all missing entries some entries were added? You will replace those entries by your fetched ones

Reply:
Actually `computeIfAbsent` should prevent that.
Looks ok, no worries

Reply:
The fetch phase and cache insert are not atomic, but the remapping of the values in the cache are. As a consequence, there could be redundant fetches, but not redundant cache remapping.

Reply:
can we use `readCache .get(...)` here?

Reply:
Using interfaces that use a loader forces you to handle execution exceptions, since we rethrow those exceptions anyways and we manage our own "loading" I don't see a strong reason to use it.